File: client/protocol_v2.js
"use strict";

const ALLOWED_TYPES = new Set(["MSG", "ACK", "NACK", "ERROR"]);

class FrameError extends Error {}
class ChecksumError extends FrameError {}

function crc32(bytes) {
  let crc = 0xffffffff;
  for (const byte of bytes) {
    crc ^= byte;
    for (let bit = 0; bit < 8; bit += 1) {
      crc = (crc >>> 1) ^ (crc & 1 ? 0xedb88320 : 0);
    }
  }
  return ((crc ^ 0xffffffff) >>> 0).toString(16).toUpperCase().padStart(8, "0");
}

function encodeFrame(type, sequence, payload) {
  if (!ALLOWED_TYPES.has(type)) throw new FrameError(`unsupported type: ${type}`);
  if (!Number.isInteger(sequence) || sequence < 0) throw new FrameError("sequence must be a non-negative integer");

  const payloadBytes = Buffer.from(payload, "utf8");
  const header = `V2|${type}|${sequence}|${payloadBytes.length}|${crc32(payloadBytes)}|`;
  return Buffer.concat([Buffer.from(header, "ascii"), payloadBytes]);
}

function decodeFrame(raw) {
  const bytes = Buffer.from(raw);
  const separators = [];
  for (let index = 0; index < bytes.length && separators.length < 5; index += 1) {
    if (bytes[index] === 124) separators.push(index);
  }
  if (separators.length !== 5) throw new FrameError("frame must contain six fields");

  const headerField = (start, end) => bytes.subarray(start, end).toString("ascii");
  const version = headerField(0, separators[0]);
  const type = headerField(separators[0] + 1, separators[1]);
  const sequence = Number(headerField(separators[1] + 1, separators[2]));
  const declaredLength = Number(headerField(separators[2] + 1, separators[3]));
  const declaredCrc = headerField(separators[3] + 1, separators[4]).toUpperCase();
  const payloadBytes = bytes.subarray(separators[4] + 1);

  if (version !== "V2") throw new FrameError(`unsupported version: ${version}`);
  if (!ALLOWED_TYPES.has(type)) throw new FrameError(`unsupported type: ${type}`);
  if (!Number.isInteger(sequence) || sequence < 0 || !Number.isInteger(declaredLength) || declaredLength < 0) {
    throw new FrameError("sequence and length must be non-negative integers");
  }
  if (!/^[0-9A-F]{8}$/.test(declaredCrc)) throw new FrameError("CRC32 must contain eight hexadecimal characters");
  if (payloadBytes.length !== declaredLength) throw new FrameError("declared payload length does not match datagram");
  if (crc32(payloadBytes) !== declaredCrc) throw new ChecksumError("payload CRC32 mismatch");

  return { version, type, sequence, payload: payloadBytes.toString("utf8") };
}

module.exports = { ChecksumError, FrameError, crc32, decodeFrame, encodeFrame };

File: server/protocol_v2.py
"""Robust V2 frame codec with CRC32 integrity verification."""

from __future__ import annotations

import binascii


ALLOWED_TYPES = {"MSG", "ACK", "NACK", "ERROR"}


class FrameError(ValueError):
    """Raised when a datagram does not match the V2 frame format."""

    def __init__(self, message: str, sequence: int | None = None) -> None:
        super().__init__(message)
        self.sequence = sequence


class ChecksumError(FrameError):
    """Raised when the payload CRC32 does not match the header."""


def _crc32(payload: bytes) -> str:
    return f"{binascii.crc32(payload) & 0xFFFFFFFF:08X}"


def encode_frame(message_type: str, sequence: int, payload: str) -> bytes:
    if message_type not in ALLOWED_TYPES:
        raise FrameError(f"unsupported type: {message_type}", sequence)
    if not isinstance(sequence, int) or sequence < 0:
        raise FrameError("sequence must be a non-negative integer")

    payload_bytes = payload.encode("utf-8")
    header = f"V2|{message_type}|{sequence}|{len(payload_bytes)}|{_crc32(payload_bytes)}|"
    return header.encode("ascii") + payload_bytes


def decode_frame(raw: bytes) -> dict[str, object]:
    fields = raw.split(b"|", 5)
    if len(fields) != 6:
        raise FrameError("frame must contain six fields")

    version_raw, type_raw, sequence_raw, length_raw, crc_raw, payload_bytes = fields
    sequence: int | None = None
    try:
        version = version_raw.decode("ascii")
        message_type = type_raw.decode("ascii")
        sequence = int(sequence_raw.decode("ascii"))
        declared_length = int(length_raw.decode("ascii"))
        declared_crc = crc_raw.decode("ascii").upper()
    except (UnicodeDecodeError, ValueError) as error:
        raise FrameError("frame header is invalid", sequence) from error

    if version != "V2":
        raise FrameError(f"unsupported version: {version}", sequence)
    if message_type not in ALLOWED_TYPES:
        raise FrameError(f"unsupported type: {message_type}", sequence)
    if sequence < 0 or declared_length < 0:
        raise FrameError("sequence and length must be non-negative", sequence)
    if len(declared_crc) != 8 or any(char not in "0123456789ABCDEF" for char in declared_crc):
        raise FrameError("CRC32 must contain eight hexadecimal characters", sequence)
    if declared_length != len(payload_bytes):
        raise FrameError("declared payload length does not match datagram", sequence)
    if declared_crc != _crc32(payload_bytes):
        raise ChecksumError("payload CRC32 mismatch", sequence)

    try:
        payload = payload_bytes.decode("utf-8")
    except UnicodeDecodeError as error:
        raise FrameError("payload is not valid UTF-8", sequence) from error

    return {"version": version, "type": message_type, "sequence": sequence, "payload": payload}
